### PR TITLE
Transposed Convoluation, ONNX export, Refactoring & Unit Tests

### PR DESCRIPTION
**Pull request by timoataltavo**
This PR adds the support of Causal Transposed Convolutions for specific parameter ranges. It also contains a refactoring the convolutions into a separate package that is extensively unit tested.

The change also helps with exporting the streamable causal convolutions as ONNX models which requires explicit handling of internal buffers via input / output tensors.

### CODE REVIEW

File: pytorch_tcn/conv.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        self.register_buffer(
+            'buffer',
+            buffer,
+            )
+        
+        return
+
+    def _forward(self, x):
+        if self.causal:
+            p = nn.ConstantPad1d(
```
Comment (paul-krug) [inline pytorch_tcn/conv.py:209]: Move to constructor as well

File: pytorch_tcn/tcn.py [hunk 2]
```diff
@@ -129,161 +131,60 @@ def get_kernel_init_fn(
...
-        x = p(x)
-        x = super().forward(x)
-        return x
-
-    def forward(
+    def inference(
             self,
             x,
-            inference=False,
+            in_buffers=None,
```
Comment (paul-krug) [inline pytorch_tcn/tcn.py:144]: should be:

def inference( *args, **kwargs ):
x = self( args, kwargs, inference=True )
return x

File: pytorch_tcn/tcn.py [hunk 3]
```diff
@@ -129,161 +131,60 @@ def get_kernel_init_fn(
...
-                Input time dimension {x.shape[2]} is too short for causal
-                inference with lookahead {self.lookahead}. You must pass at
-                least lookhead + 1 time steps ({self.lookahead + 1}).
-                """
-                )
-        x = torch.cat(
-            (self.buffer, x),
-            -1,
+        
+        return self.forward(
```
Comment (paul-krug) [inline pytorch_tcn/tcn.py:148]: see comment above, self.forward() ignores possible hooks

File: pytorch_tcn/tcn.py [hunk 4]
```diff
@@ -547,9 +420,9 @@ def __init__(
             embedding_shapes: Optional[ ArrayLike ] = None,
             embedding_mode: str = 'add',
             use_gate: bool = False,
-            lookahead: int = 0,
```
Comment (paul-krug) [inline pytorch_tcn/tcn.py:550]: ordering of parameters should stay the same

File: setup.py [hunk 5]
```diff
@@ -3,6 +3,8 @@
 install_requires = [
     'torch',
     'numpy',
+    'onnx',
```
Comment (paul-krug) [inline setup.py:6]: Better to install onnx packages only via the CI config for unit testing, but do not make it a requirement for the package

**Review by paul-krug — CHANGES_REQUESTED** [review]
Looks great overall, but some minor issues should be addressed

File: .vscode/settings.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    "python.testing.unittestEnabled": true,
+    "python.testing.pytestEnabled": false,
+    "python.testing.unittestArgs": [
+        "-v",
+        "-s",
+        "./tests",
+        "-p",
+        "test*.py"
+    ]
+}
```
[the remark this answers is not in the document]
Comment (timoataltavo) [reply .vscode/settings.json]: Agreed. Removed it and added .vscode to the .gitignores.

File: pytorch_tcn/conv.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        # Register buffer as a persistent buffer which is available as self.buffer
+        self.register_buffer(
+            'buffer',
+            buffer,
+            )
+        
+        return
+    
+    def _forward(self, x):
+        p = nn.ConstantPad1d(
```
[the remark this answers is not in the document]
Comment (timoataltavo) [reply pytorch_tcn/conv.py]: Done

File: pytorch_tcn/conv.py (hunk 1 above)
Comment (timoataltavo) [reply pytorch_tcn/conv.py]: Done

File: pytorch_tcn/tcn.py (hunk 2 above)
Comment (timoataltavo) [reply pytorch_tcn/tcn.py]: Done.

File: pytorch_tcn/tcn.py (hunk 3 above)
Comment (timoataltavo) [reply pytorch_tcn/tcn.py]: Done.

File: pytorch_tcn/tcn.py (hunk 4 above)
Comment (timoataltavo) [reply pytorch_tcn/tcn.py:550]: Fixed.

File: setup.py (hunk 5 above)
Comment (timoataltavo) [reply setup.py]: Onnx is now optional and tests that require it will fail gracefully

**Review by paul-krug — APPROVED** [review]

**Pull request merged**